File: WINGS_AI_Bootcamp/Elixir/checkPowerOf2.ex
import Bitwise


defmodule PowerOf2 do
  # Using bitwise AND (like first Python function)
  def is_power2(n) when n > 0 do
    (n &&& (n - 1)) == 0
  end
  def is_power2(_), do: false

  # Recursive check (like second Python function)
  def isp2(0), do: false
  def isp2(1), do: true
  def isp2(n) when rem(n, 2) == 1, do: false
  def isp2(n), do: isp2(div(n, 2))
end

# Example test runs
IO.inspect PowerOf2.is_power2(16)  # true
IO.inspect PowerOf2.is_power2(18)  # false

IO.inspect PowerOf2.isp2(16)       # true
IO.inspect PowerOf2.isp2(18)       # false

File: WINGS_AI_Bootcamp/Haskell/checkPowerOf2.hs
import Data.Bits ((.&.))

-- Bitwise AND method
isPower2 :: Int -> Bool
isPower2 n = n > 0 && (n .&. (n - 1)) == 0

-- Recursive method
isp2 :: Int -> Bool
isp2 0 = False
isp2 1 = True
isp2 n
  | odd n     = False
  | otherwise = isp2 (n `div` 2)

main :: IO ()
main = do
  print (isPower2 16)  -- True
  print (isPower2 18)  -- False
  print (isp2 16)      -- True
  print (isp2 18)      -- False

